src/samples/bluetooth/simple-pair.c

```diff
@@ -0,0 +1,184 @@
+{
+    const struct sol_bt_device_info *info = sol_bt_conn_get_device_info(conn);
+
+    SOL_DBG("Paired with %s", info->name);
+}
+
+static bool
+on_connect(void *user_data, struct sol_bt_conn *conn)
+{
+    SOL_BUFFER_DECLARE_STATIC(str, SOL_BLUETOOTH_ADDR_STRLEN);
+    int r;
+
+    sol_network_link_addr_to_str(sol_bt_conn_get_addr(conn), &str);
+
+    SOL_INF("Connected to device %.*s", (int)str.used, (char *)str.data);
+
+    r = sol_bt_conn_pair(conn, paired_callback, NULL);
+    SOL_INT_CHECK(r, < 0, false);
+
+    return true;
+}
+
+static void
+on_disconnect(void *user_data, struct sol_bt_conn *conn)
+{
+    SOL_BUFFER_DECLARE_STATIC(str, SOL_BLUETOOTH_ADDR_STRLEN);
+
+    sol_network_link_addr_to_str(sol_bt_conn_get_addr(conn), &str);
+
+    SOL_INF("Disconnected from device %.*s", (int)str.used, (char *)str.data);
+
+    auth_conn = NULL;
+}
+
+static bool
+timeout_cb(void *data)
+{
+    auth_conn = sol_bt_connect(&pair_addr, on_connect, on_disconnect,
+        on_error, NULL);
+    SOL_NULL_CHECK(auth_conn, false);
+
+    return false;
+}
+
+static void
+found_device(void *user_data, const struct sol_bt_device_info *device)
+{
+    SOL_BUFFER_DECLARE_STATIC(str, SOL_BLUETOOTH_ADDR_STRLEN);
+    const char *addr;
+
+    addr = sol_network_link_addr_to_str(&device->addr, &str);
+    SOL_NULL_CHECK(addr);
+
+    SOL_INF("device %.*s in range %s", (int)str.used, (char *)str.data,
+        device->in_range ? "yes" : "no");
+
+    if (pair_addr.family != SOL_NETWORK_FAMILY_UNSPEC) {
+        if (!sol_network_link_addr_eq(&pair_addr, &device->addr))
+            return;
+
+        sol_bt_stop_scan(scan);
+        scan = NULL;
+
+        timeout = sol_timeout_add(500, timeout_cb, NULL);
+    }
+}
+
+static void
+pairing_confirm(void *data, struct sol_bt_conn *conn)
+{
+    SOL_DBG("conn %p auth_conn %p", conn, auth_conn);
+
+    if (auth_conn != conn) {
+        sol_bt_agent_reply_cancel(conn);
+        return;
+    }
+
+    sol_bt_agent_reply_pairing_confirm(conn);
+}
+
+static struct sol_bt_agent agent = {
+    .pairing_confirm = pairing_confirm,
+};
+
+static void
+enabled(void *data, bool powered)
+{
+    int r;
+
+    if (!powered)
+        return;
+
+    SOL_INF("Bluetooth Adapter enabled");
+
+    scan = sol_bt_start_scan(SOL_BT_TRANSPORT_ALL, found_device, NULL);
+    SOL_NULL_CHECK(scan);
+
+    r = sol_bt_register_agent(&agent, NULL);
+    if (r < 0) {
+        SOL_WRN("r (%d) < 0", r);
+        sol_quit_with_code(r);
+    }
+}
+
+static void
+shutdown(void)
+{
+    if (auth_conn)
+        sol_bt_conn_unref(auth_conn);
+
+    if (scan)
+        sol_bt_stop_scan(scan);
+
+    if (session)
+        sol_bt_disable(session);
+
+    sol_bt_unregister_agent(&agent);
+}
+
+static void
+startup(void)
+{
+    if (sol_argc() > 1) {
+        const struct sol_network_link_addr *addr;
+
+        addr = sol_network_link_addr_from_str(&pair_addr, sol_argv()[1]);
+        SOL_NULL_CHECK(addr);
```

Comment:
this would not quit the application when an invalid address is used, is that ok?